grpc.go

```diff
@@ -0,0 +1,698 @@
+
+		proc.process.ConsoleSocket = childSock
+		proc.consoleSock = parentSock
+
+		return proc, nil
+	}
+
+	rStdin, wStdin, err := os.Pipe()
+	if err != nil {
+		return nil, err
+	}
+
+	rStdout, wStdout, err := os.Pipe()
+	if err != nil {
+		return nil, err
+	}
+
+	rStderr, wStderr, err := os.Pipe()
+	if err != nil {
+		return nil, err
+	}
+
+	proc.process.Stdin = rStdin
+	proc.process.Stdout = wStdout
+	proc.process.Stderr = wStderr
+
+	proc.stdin = wStdin
+	proc.stdout = rStdout
+	proc.stderr = rStderr
+
+	return proc, nil
+}
+
+// Shared function between StartContainer and ExecProcess, because those expect
+// a process to be run. The difference being the process does not exist yet in
+// case of ExecProcess.
+func (a *agentGRPC) runProcess(cid string, agentProcess *pb.Process) (err error) {
+	if a.pod.running == false {
+		return fmt.Errorf("Pod not started")
+	}
+
+	ctr, err := a.pod.getContainer(cid)
+	if err != nil {
+		return err
+	}
+
+	status, err := ctr.container.Status()
+	if err != nil {
+		return err
+	}
+
+	var proc *process
+	if agentProcess != nil {
+		if status != libcontainer.Running {
+			return fmt.Errorf("Container %s status %s, should be %s", cid, status.String(), libcontainer.Running.String())
+		}
+
+		if _, err := ctr.getProcess(agentProcess.Id); err == nil {
+			return fmt.Errorf("Process %s already exists", agentProcess.Id)
+		}
+
+		proc, err = buildProcess(agentProcess)
+		if err != nil {
+			return err
+		}
+	} else {
+		if status != libcontainer.Created {
+			return fmt.Errorf("Container %s status %s, should be %s", cid, status.String(), libcontainer.Created.String())
+		}
+
+		proc, err = ctr.getProcess(ctr.initProcId);
+		if err != nil {
+			return err
+		}
+	}
+
+	ctr.wgProcesses.Add(1)
+
+	if err := ctr.container.Run(&(proc.process)); err != nil {
+		return fmt.Errorf("Could not run process: %v", err)
+	}
+	defer proc.closePostStartFDs()
+
+	// Setup terminal if enabled.
+	if proc.consoleSock != nil {
+		termMaster, err := utils.RecvFd(proc.consoleSock)
+		if err != nil {
+			return err
+		}
+
+		if err := setConsoleCarriageReturn(int(termMaster.Fd())); err != nil {
+			return err
+		}
+
+		proc.termMaster = termMaster
+	}
+
+	// Update process info.
+	ctr.setProcess(proc.id, proc)
+
+	return nil
+}
+
+func (a *agentGRPC) CreateContainer(ctx context.Context, req *pb.CreateContainerRequest) (*gpb.Empty, error) {
+	if a.pod.running == false {
+		return emptyResp, fmt.Errorf("Pod not started, impossible to run a new container")
+	}
+
+	if _, err := a.pod.getContainer(req.Container.Id); err == nil {
+		return emptyResp, fmt.Errorf("Container %s already exists, impossible to create", req.Container.Id)
+	}
+
+	// re-scan PCI bus
+	// looking for hidden devices
+	if err := ioutil.WriteFile(pciBusRescanFile, []byte("1"), pciBusMode); err != nil {
+		agentLog.WithError(err).Warnf("Could not rescan PCI bus")
+	}
+
+	defaultMountFlags := syscall.MS_NOEXEC | syscall.MS_NOSUID | syscall.MS_NODEV
+
+	config := configs.Config{
```

Comment:
all the configs here should be replaced by the config from the request, or merged from the request.

Reply:
Agreed with @laijs here. You probably want to convert the `req.OCI` structure into an actual `oci.OCI` and use as much as possible from it.

Reply:
@sameo @lajis, if I understand correctly, you mean that we should expect everything to be provide by `req.OCI` ? Are you sure we don't need some default values in the agent ?

Reply:
IMO, agent can have its default values for the container, but the default value can be overwritten by the requested values. And runtime is also encouraged to pass more config even the agent has default value. For example runtime is encouraged to pass the procfs setting of the container into the agent.

Reply:
@laijs Makes sense, I'll do that.